utils/fswriters.go

```diff
@@ -0,0 +1,66 @@
+package utils
+
+import (
+	"io"
+	"os"
+)
+
+type atomicFileWriter struct {
+	f        *os.File
+	path     string
+	writeErr error
+	perm     os.FileMode
+}
+
+func (a *atomicFileWriter) write(p []byte) (n int, err error) {
+	n, err = a.f.Write(p)
+	if err != nil {
+		a.writeErr = err
+	}
+	return n, err
+}
+
+func (a *atomicFileWriter) close() (err error) {
+	defer func() {
+		if err != nil || a.writeErr != nil {
+			CleanFile(a.f)
+		}
+	}()
+	if err = a.f.Sync(); err != nil {
+		a.f.Close()
+		return err
+	}
+	if err := a.f.Close(); err != nil {
+		return err
+	}
+	if err := os.Chmod(a.f.Name(), a.perm); err != nil {
+		return err
+	}
+	if a.writeErr == nil {
+		return os.Rename(a.f.Name(), a.path)
```

Comment:
What could happen if a.path exists already? os.Rename can overwrite it?
return err or overwrite the file? which is better?

Reply:
If newpath already exists and is not a directory, Rename replaces it

Reply:
If my target is "/test/sub/file", but dir "/test/sub" not exists, what will happen?